src/plot_api/plot_api.js

```diff
@@ -2477,6 +2488,249 @@ Plotly.relayout = function relayout(gd, astr, val) {
+                continue;
+            }
+
+            transitioningTraces.push(traceIdx);
+
+            newTraceData = newData[i];
+            curData = gd.data[traces[i]];
+
+            for(var ai in newTraceData) {
+                var value = newTraceData[ai];
+                Lib.nestedProperty(curData, ai).set(value);
+            }
+
+            var traceIdx = traces[i];
+            if(gd.data[traceIdx].marker && gd.data[traceIdx].marker.size) {
+                gd._fullData[traceIdx].marker.size = gd.data[traceIdx].marker.size;
+            }
+            if(gd.data[traceIdx].error_y && gd.data[traceIdx].error_y.array) {
+                gd._fullData[traceIdx].error_y.array = gd.data[traceIdx].error_y.array;
+            }
+            if(gd.data[traceIdx].error_x && gd.data[traceIdx].error_x.array) {
+                gd._fullData[traceIdx].error_x.array = gd.data[traceIdx].error_x.array;
+            }
+            gd._fullData[traceIdx].x = gd.data[traceIdx].x;
+            gd._fullData[traceIdx].y = gd.data[traceIdx].y;
+            gd._fullData[traceIdx].z = gd.data[traceIdx].z;
+            gd._fullData[traceIdx].key = gd.data[traceIdx].key;
+        }
+
+        doCalcdata(gd, transitioningTraces);
+
+        ErrorBars.calc(gd);
+    }
+
+    function doAnimations() {
+        var a, i, j;
+        var basePlotModules = fullLayout._basePlotModules;
+        for(j = 0; j < basePlotModules.length; j++) {
+            basePlotModules[j].plot(gd, transitioningTraces, transitionOpts);
+        }
+
+        if(layout) {
+            for(j = 0; j < basePlotModules.length; j++) {
+                if(basePlotModules[j].transitionAxes) {
+                    basePlotModules[j].transitionAxes(gd, layout, transitionOpts);
+                }
+            }
+        }
+
+        if(!transitionOpts.leadingEdgeRestyle) {
+            return new Promise(function(resolve, reject) {
+                completion = resolve;
+                completionTimeout = setTimeout(resolve, transitionOpts.duration);
+            });
+        }
+    }*/
+};
+
+/**
+ * Animate to a keyframe
+ *
+ * @param {string} name
+ *      name of the keyframe to create
+ * @param {object} transitionConfig
+ *      configuration for transition
+ */
+Plotly.animate = function(gd, name /*, transitionConfig*/) {
+    gd = getGraphDiv(gd);
+
+    var _frames = gd._frameData._frames;
+
+    if(!_frames[name]) {
+        Lib.warn('animateToFrame failure: keyframe does not exist', name);
+        return Promise.reject();
+    }
+
+    return Promise.resolve();
+};
+
+/**
+ * Create new keyframes
+ *
+ * @param {array of objects} frameList
+ *      list of frame definitions, in which each object includes any of:
+ *      - name: {string} name of keyframe to add
+ *      - data: {array of objects} trace data
+ *      - layout {object} layout definition
+ *      - traces {array} trace indices
+ *      - baseFrame {string} name of keyframe from which this keyframe gets defaults
```

Comment:
To my :eyes: , I think the later makes the most sense.

I like to think of frames as diffs for `data` and `layout`. How to transition between these diffs should be a property of the widget (e.g. slider, dropdown menus). For example, given a set of frames `a`, `b`, `c`, then:




```js
var layout = {
  slider: {
    frames: ['a', 'b', 'c'],
    transition: {
      delay: 100,
      duration: 1000,
      easing: 'cubic'
    }
  }
};
```

and per-frame transition options would then be set as:

```js
var layout = {
  slider: {
    frames: ['a', 'b', 'c'],
    transition: {
      delay: [100, 200, 300],
      duration: [1000, 2000, 3000],
      easing: ['cubic', 'linear', 'sin']
    }
  }
};
```